File: R/Descriptive Analytics/ConductFactorAnalysis.R
# Load packages
library(dplyr)
library(factoextra)
library(nFactors)
library(psych)

# Declare function
ConductFactorAnalysis = function(dataframe,
                                 list_of_variables = NULL,
                                 number_of_factors_to_create = NULL,
                                 measure_of_sampling_adequacy_threshold = 0.5,
                                 pca_standard_deviation_threshold = 1,
                                 standardize_variables = TRUE,
                                 plot_pca_results = TRUE) {
  # Select variables and keep complete cases only
  if (is.null(list_of_variables)) {
    dataframe_kmo = dataframe %>% 
      dplyr::select_if(is.numeric) %>% 
      na.omit
  } else {
    dataframe_kmo = dataframe %>% 
      select(
        all_of(list_of_variables)
      ) %>%
      na.omit
  }
  
  # Run KMO test
  kmo_results = KMO(dataframe_kmo)
  print(kmo_results)
  message("The KMO (Kaiser-Meyer-Olkin) test aims to estimate the proportion of variance among the variables you selected that may be caused by some underlying factor(s).")
  message("The closer the Overall Measure of Sampling Adequacy (MSA) is to 1, the more useful factor analysis will be for your selected variables.")
  
  # See if KMO test result is below threshold 
  if (kmo_results$MSA < measure_of_sampling_adequacy_threshold) {
    warning(paste0("The Overall MSA is below your threshold of ", measure_of_sampling_adequacy_threshold, ". Factor analysis is not recommended!"))
  } else {
    message(paste("The Overall MSA is above your threshold of", measure_of_sampling_adequacy_threshold))
  }
  
  # Get list of variables that are above MSA threshold
  list_of_pca_eligible_vars = as.data.frame(kmo_results$MSAi)
  names(list_of_pca_eligible_vars)[names(list_of_pca_eligible_vars) == "kmo_results$MSAi"] = "MSA"
  list_of_pca_eligible_vars = list_of_pca_eligible_vars %>%
    filter(
      MSA >= measure_of_sampling_adequacy_threshold
    )
  list_of_pca_eligible_vars = c(row.names(list_of_pca_eligible_vars))
  
  # Select eligible variables and keep complete cases only
  dataframe_pca = dataframe %>% 
    select(all_of(list_of_pca_eligible_vars)) %>% 
    na.omit
  
  # Conduct PCA analysis
  pca_model = princomp(dataframe_pca,
                       cor = standardize_variables,
                       scores = TRUE)
  
  # If requested, show graph of variables
  if (plot_pca_results) {
    p = fviz_pca_var(
      pca_model,
      col.var = "contrib",
      gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
      repel = TRUE
    )
    plot(p)
  }
  
  # Get number of factors to use based on st. dev., or eigenvalue, threshold
  pca_sdev = as.data.frame(pca_model$sdev)
  names(pca_sdev)[names(pca_sdev) == "pca_model$sdev"] = "sdev"
  pca_sdev = pca_sdev %>% 
    filter(
      sdev >= 1
    )
  
  # Conduct final factor analysis
  factor_analysis_results = factanal(dataframe_pca, 
                                     nrow(pca_sdev), 
                                     rotation="varimax")
  
  # Show factor analysis results
  print(factor_analysis_results, 
        digits=2, 
        cutoff=.3, 
        sort=TRUE)
  
  # Select only factor-relevant columns
  dataframe_factors = dataframe %>% 
    select(all_of(list_of_pca_eligible_vars))
  
  # Add factors to original dataset
  df_factors = factor.scores(dataframe_factors, factor_analysis_results)
  df_factors = as.data.frame(df_factors$scores)
  dataframe = bind_cols(
    dataframe,
    df_factors
  )
  
  # Return dataframe
  return(dataframe)
}

File: Python/Descriptive Analytics/ConductFactorAnalysis.py
# Load packages 
from factor_analyzer.factor_analyzer import calculate_bartlett_sphericity
from factor_analyzer.factor_analyzer import calculate_kmo
from factor_analyzer import FactorAnalyzer
from logging import warning
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import StandardScaler

# Declare function
def ConductFactorAnalysis(dataframe,
                          list_of_variables = None,
                          standardize_variables = True,
                          number_of_factors_to_create = None,
                          bartlett_sphericity_significance_level = 0.05,
                          measure_of_sampling_adequacy_threshold = 0.6,
                          pca_eigenvalue_threshold = 1):
    """_summary_
    This function conducts factor analysis on a dataset.

    Args:
        dataframe (Pandas dataframe): Pandas dataframe containing the data to be analyzed.
        list_of_variables (list, optional): _description_. The list of variables to be analyzed. Defaults to None, which means all variables will be analyzed.
        standardize_variables (bool, optional): _description_. Whether to standardize the variables prior to analysis. Defaults to True.
        number_of_factors_to_create (int, optional): _description_. The number of factors to create. Defaults to None, which means the number of factors will be determined automatically.
        bartlett_sphericity_significance_level (float, optional): _description_. The significance level for the Bartlett test for sphericity. Defaults to 0.05.
        measure_of_sampling_adequacy_threshold (float, optional): _description_. The threshold for the Keiser-Meyer-Olkin (KMO) test. Defaults to 0.6.
        pca_eigenvalue_threshold (int, optional): _description_. The threshold for the eigenvalue in the principal component analysis (PCA). Defaults to 1.
        
    Returns:
        Pandas dataframe: An updated Pandas dataframe with factor analysis results joined to the original dataset.
    """
    
    # Select list of variables to consider
    if list_of_variables == None:
        list_of_variables = dataframe.columns
        df_for_testing = dataframe.copy()
    else:
        df_for_testing = dataframe[list_of_variables].copy()
    
    # Filter to complete cases only
    df_for_testing = df_for_testing.dropna()
    
    # Run Bartlett test
    print("The Bartlett test for sphericity helps to determine whether there is a statistically detectable redundancy between the variables you selected. If so, those variables can be summarized with a few(er) number of factors.")
    bartlett_chi_square_value, bartless_p_value = calculate_bartlett_sphericity(df_for_testing)
    
    # See if Barlett test p-value is statistically significant. Show a warning if it is not.
    if bartless_p_value > bartlett_sphericity_significance_level:
        warning("The Bartlett test was not statistically significant at the " + str(bartlett_sphericity_significance_level) + " level -- meaning there is not a detectable redundancy among the variables you selected. Factor analysis is not recommended!")
    else:
        print("The Bartlett test was statistically significant at the " + str(bartlett_sphericity_significance_level) + " level -- meaning there is a detectable redundancy among the variables you selected.")
    
    # Run Keiser-Meyer-Olkin (KMO) test
    print("\n\nThe Keiser-Meyer-Olkin (KMO) test aims to estimate the proportion of variance among the variables you selected that may be explained by some underlying factor(s). The close the overall measure of sampling adequacy (MSA) is to 1, the more useful factor analysis will be in reducing your variables to a few factors.")
    kmo_all, kmo_model = calculate_kmo(df_for_testing)
    
    # See if KMO test result is below measure of sampling adequacy threshold
    if kmo_model < measure_of_sampling_adequacy_threshold:
        warning("The overall MSA is " + str(round(kmo_model, 3)) + ", which is below your desired threshold of " + str(measure_of_sampling_adequacy_threshold) + ". Factor analysis is not recommended!")
    else:
        print("The overall MSA is " + str(round(kmo_model, 3)) + ", which meets or exceeds your desired threshold of " + str(measure_of_sampling_adequacy_threshold) + ".")
    
    # Get list of variables that are above measure of sampling adequacy threshold
    df_variables = pd.DataFrame({
        'Variable': list_of_variables,
        'MSA': kmo_all
    })
    df_variables = df_variables[
        df_variables['MSA'] >= measure_of_sampling_adequacy_threshold
    ]
    list_of_eligible_variables = df_variables['Variable'].to_list()
    
    # Select variables that are above measure of sampling adequacy threshold
    df_for_fa = dataframe[list_of_eligible_variables].copy()
    
    # Filter to complete cases only
    df_for_fa = df_for_fa.dropna()
    
    # Standardize columns if requested
    if standardize_variables:
        df_transformed = StandardScaler().fit_transform(df_for_fa)
        df_transformed = pd.DataFrame(df_transformed)
        df_transformed.index = df_for_fa.index
        df_for_fa = df_transformed.copy()
        df_for_fa.columns = list_of_eligible_variables
    
    # Conduct principal component analysis
    fa = FactorAnalyzer()
    fa.fit(df_for_fa)
    
    # TO-DO: Show graph of variables
    
    # Get number of factors to use based on eigenvalue threshold (standard deviation)
    if number_of_factors_to_create == None:
        ev, v = fa.get_eigenvalues()
        number_of_factors_to_create = len(ev[ev >= pca_eigenvalue_threshold])
    
    # Conduct final factor analysis
    fa = FactorAnalyzer(
        n_factors=number_of_factors_to_create,
        rotation="verimax",
        method="minres",
        use_smc=True
    )
    fa.fit(df_for_fa)
    
    # Create list of factor names
    list_factor_names = []
    for i in range(0, number_of_factors_to_create):
        list_factor_names.append("Factor " + str(i + 1))
    
    # Show variance of each factor
    pd.DataFrame(
        fa.get_factor_variance(),
        columns=list_factor_names,
        index=["SS Loadings", "Share of Variance Explained (%)", "Cumulative Variance Explained"]
    )
    
    # Select variables that are above measure of sampling adequacy threshold
    df_factors = dataframe[list_of_eligible_variables].copy()
    
    # Filter to complete cases only
    df_factors = df_factors.dropna()
    
    # Predict factors for observed values in original dataset
    if standardize_variables:
        df_transformed = StandardScaler().fit_transform(df_factors)
        df_transformed = pd.DataFrame(df_transformed)
        df_transformed.index = df_factors.index
        df_factors = df_transformed.copy()
        df_factors.columns = list_of_eligible_variables
    df_predicted = fa.transform(df_factors)
    df_predicted = pd.DataFrame(
        df_predicted,
        columns=list_factor_names,
        index=df_factors.index
    )
    
    # Join predicted/fitted factors to original dataset
    dataframe = dataframe.merge(
        df_predicted,
        how='left',
        left_index=True,
        right_index=True
    )
    
    # Return dataset with factors added
    return(dataframe)
